lib/puppet/provider/mongodb_replset/mongo.rb

```diff
@@ -201,12 +232,33 @@ def set_members
       if output['ok'] == 0
         raise Puppet::Error, "rs.initiate() failed for replicaset #{self.name}: #{output['errmsg']}"
       end
+
+      # Check that the replicaset has finished initialization
+      retry_limit = 10
+      retry_sleep = 3
+
+      retry_limit.times do |n|
+        begin
+          if db_ismaster(alive_hosts[0])['ismaster']
+            Puppet.debug 'Replica set initialization has successfully ended'
+            return
+          else
+            Puppet.debug "Wainting for replica initialization. Retry: #{n}"
+            sleep retry_sleep
+            next
+          end
+        end
+      end
+      raise Puppet::Error, "rs.initiate() failed for replicaset #{self.name}: host #{alive_hosts[0]} didn't become master"
+
     else
       # Add members to an existing replset
+      Puppet.debug "Adding member to existing replset #{self.name}"
       if master = master_host(alive_hosts)
         current_hosts = db_ismaster(master)['hosts']
```

Comment:
@fatmcgav: This should get arbiters as well or else it will try to add the arbiter even though it is already added.

I will try to prepare a patch for this.

```
services-lxstage:PRIMARY> db.isMaster()
{
	"setName" : "services-XXX",
	"setVersion" : 1,
	"ismaster" : true,
	"secondary" : false,
	"hosts" : [
		"services-mongodb-02.XX:27017"
	],
	"arbiters" : [
		"services-mongodb-00.XX:27017"
	],
	"primary" : "services-mongodb-02.XX:27017",
	"me" : "services-mongodb-02.XX:27017",
	"electionId" : ObjectId("5604f2a2c3f732daff9b4d6c"),
	"maxBsonObjectSize" : 16777216,
	"maxMessageSizeBytes" : 48000000,
	"maxWriteBatchSize" : 1000,
	"localTime" : ISODate("2015-09-25T08:07:01.145Z"),
	"maxWireVersion" : 2,
	"minWireVersion" : 0,
	"ok" : 1
}
```